File: backend/schemas/memory.py
"""Pydantic models for memory management APIs.

This module defines the ``Memory`` schema used by the `/memory` API.
The schema mirrors the fields defined on the ``Memory`` ORM model to
expose persisted memories via FastAPI while enforcing type safety.  When
the ORM model is extended with new fields, update this schema
accordingly and regenerate the frontend TypeScript definitions to avoid
misalignment:contentReference[oaicite:4]{index=4}.
"""
from __future__ import annotations

from datetime import datetime
from typing import Optional, Dict, Any
from pydantic import BaseModel


class Memory(BaseModel):
    """Serialized view of a memory entry."""

    id: int
    tenant_id: str
    project_id: Optional[str] = None
    kind: str
    created_at: datetime
    tags: Optional[Dict[str, Any]] = None
    trace_id: Optional[str] = None
    sha256: Optional[str] = None
    prev_sha256: Optional[str] = None

    model_config = {"from_attributes": True}

File: frontend/src/types/memory.ts
/**
 * Type definitions for memory entries returned by the API.
 *
 * This interface mirrors the server‑side `Memory` Pydantic schema:contentReference[oaicite:8]{index=8}.
 * Whenever the backend `Memory` model changes (e.g. a new property is added),
 * update this type as well as the Pydantic schema to keep all layers aligned.
 * Optional properties such as `project_id` and `tags` are represented
 * accordingly. `created_at` is an ISO timestamp string.
 */
export interface Memory {
  id: number;
  tenant_id: string;
  project_id?: string | null;
  kind: string;
  created_at: string;
  tags?: Record<string, any> | null;
  trace_id?: string | null;
  sha256?: string | null;
  prev_sha256?: string | null;
}
